### ISSUE TITLE
Unreferenced buffers in flatbuffer

### ISSUE THREAD

**Issue by Tombana** [tensorflow/tensorflow#44353]
When using `inference_input_type = tf.int8` or `inference_output_type = tf.int8`, there are unrefenrenced buffers leftover in the flatbuffer. The problem is in [this function](https://github.com/tensorflow/tensorflow/blob/b0784e587b62eec6967196b745bba4db3a90ab0c/tensorflow/lite/python/util.py#L632), which cleans up unreferenced *tensors* [over here](https://github.com/tensorflow/tensorflow/blob/b0784e587b62eec6967196b745bba4db3a90ab0c/tensorflow/lite/python/util.py#L710) but it does not cleanup unreferenced *buffers* (i.e. the `BufferT` object).

It can be reproduced with:

```python
import tensorflow as tf

def mymodel():
    img = tf.keras.layers.Input(shape=(96, 96, 3))
    x = img
    x = tf.quantization.fake_quant_with_min_max_vars(x, -3, 3)
    x = tf.keras.layers.Conv2D(32, 3)(x)
    x = tf.quantization.fake_quant_with_min_max_vars(x, -3, 3)
    return tf.keras.Model(img, x)

converter = tf.lite.TFLiteConverter.from_keras_model(mymodel())
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.inference_input_type = tf.int8
converter.inference_output_type = tf.int8
tflite_model = converter.convert()
```

In the above example, when `inference_input_type` and `inference_output_type` are not set, the flatbuffer will contain 7 `Buffer` objects, of which `Buffer` 0 is empty and unreferenced by definition of the schema. `Buffer`s 1 and 6 are referenced by the input and output tensors of the model. When `inference_input_type` and `inference_output_type` are set to `tf.int8`, then `Buffer`s 1 and 6 remain in the flatbuffer file but are unreferenced.

@MeghnaNatraj can you take a look at this? (I saw you pushed the code related to this)

**System information**
- OS Platform and Distribution: Arch Linux
- TensorFlow installed from (source or binary): binary
- TensorFlow version: `tf-nightly` on 27 Oct 2020.

**Comment by MeghnaNatraj** [comment]
Thank you for pointing this out! I'll take a look at this and get back to you. I have indeed missed cleaning up the buffers, and only remove unused tensors.

**Comment by MeghnaNatraj** [comment]
Hello @Tombana Thanks for pointing this out. I've looked into it and yes you're right..unreferenced buffers don't get cleaned up. However, we have decided to keep them as is. Reasons:

1. Buffers are associated with a model while tensors are associated with a model's subgraph. As a result, we may have tensors of different subgraphs reference the same buffers. 
2. Overhead: There are very few model input/output buffers usually, and to scan through the entire model and update all buffer references is currently not required. 

I've filed a bug internally, and we will look into this when unreferenced buffers is a priority. Will close the issue for now.
